diff --git a/KernelSU/kernel/Kconfig b/KernelSU/kernel/Kconfig
index 994ab373..47e192ed 100644
--- a/KernelSU/kernel/Kconfig
+++ b/KernelSU/kernel/Kconfig
@@ -118,6 +118,50 @@ config KSU_SUSFS_SUS_MAP
         - It may not be able to evade detections by apps that implement a good injection detection.
         - Effective only on zygote spawned umounted user app process.
 
+config KSU_SUSFS_HAS_MAGIC_MOUNT
+    bool "Say yes if the current KernelSU repo has magic mount implemented (default y)"
+    depends on KSU_SUSFS
+    default y
+    help
+        - Enable to indicate that the current SUSFS kernel supports the auto hide features for Magic Mount KernelSU.
+        - Magic Mount KernelSU (e.g. 5ec1cff) mounts modules under /debug_ramdisk/workdir and SUSFS needs
+          to be aware of it to apply the auto-hide features properly.
+
+config KSU_SUSFS_AUTO_ADD_SUS_KSU_DEFAULT_MOUNT
+    bool "Enable to hide KSU's default mounts automatically (experimental)"
+    depends on KSU_SUSFS_SUS_MOUNT
+    default y
+    help
+        - Automatically add KSU's default mounts to sus_mount.
+        - Only mount operation from process with ksu domain will be checked.
+        - Requires the underlying SUSFS kernel to ship the auto_add_sus_ksu_default_mount implementation.
+
+config KSU_SUSFS_AUTO_ADD_SUS_BIND_MOUNT
+    bool "Enable to hide suspicious bind mounts automatically (experimental)"
+    depends on KSU_SUSFS_SUS_MOUNT
+    default y
+    help
+        - Automatically add binded mounts to sus_mount.
+        - Only mount operation from process with ksu domain will be checked.
+        - Requires the underlying SUSFS kernel to ship the auto_add_sus_bind_mount implementation.
+
+config KSU_SUSFS_AUTO_ADD_TRY_UMOUNT_FOR_BIND_MOUNT
+    bool "Enable to add bind mounts to ksu's try_umount automatically (experimental)"
+    depends on KSU_SUSFS_TRY_UMOUNT
+    default n
+    help
+        - Automatically add binded mounts to ksu's try_umount.
+        - Only mount operation from process with ksu domain will be checked.
+        - Requires the underlying SUSFS kernel to ship the auto_add_try_umount_for_bind_mount implementation.
+
+config KSU_SUSFS_SUS_SU
+    bool "Enable SUS-SU in runtime temporarily"
+    depends on KSU_SUSFS
+    default n
+    help
+        - Allow user to enable or disable core ksu hooks for su temporarily.
+        - This option only makes sense when the underlying SUSFS kernel ships the sus_su support.
+
 endmenu
 
 endmenu
diff --git a/KernelSU/kernel/setuid_hook.c b/KernelSU/kernel/setuid_hook.c
index db7990ec..62d0e882 100644
--- a/KernelSU/kernel/setuid_hook.c
+++ b/KernelSU/kernel/setuid_hook.c
@@ -198,3 +198,8 @@ void ksu_setuid_hook_exit(void)
 	ksu_kernel_umount_exit();
 	ksu_unregister_feature_handler(KSU_FEATURE_ENHANCED_SECURITY);
 }
+
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+void ksu_susfs_enable_sus_su(void) { }
+void ksu_susfs_disable_sus_su(void) { }
+#endif // CONFIG_KSU_SUSFS_SUS_SU
diff --git a/KernelSU/kernel/supercalls.c b/KernelSU/kernel/supercalls.c
index ffa7383d..958a0091 100644
--- a/KernelSU/kernel/supercalls.c
+++ b/KernelSU/kernel/supercalls.c
@@ -727,6 +727,10 @@ int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd,
 			susfs_set_hide_sus_mnts_for_all_procs((void __user *)arg);
 			return 0;
 		}
+		if (cmd == CMD_SUSFS_ADD_SUS_MOUNT) {
+			susfs_add_sus_mount((void __user *)arg);
+			return 0;
+		}
 #endif //#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 #ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
 		if (cmd == CMD_SUSFS_ADD_SUS_KSTAT) {
@@ -747,6 +751,11 @@ int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd,
 			susfs_add_try_umount((void __user *)arg);
 			return 0;
 		}
+		extern void susfs_run_try_umount_for_current_mnt_ns(void);
+		if (cmd == CMD_SUSFS_RUN_UMOUNT_FOR_CURRENT_MNT_NS) {
+			susfs_run_try_umount_for_current_mnt_ns();
+			return 0;
+		}
 #endif //#ifdef CONFIG_KSU_SUSFS_TRY_UMOUNT
 #ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
 		if (cmd == CMD_SUSFS_SET_UNAME) {
